core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -84,19 +77,74 @@ public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpira
     deleteFiles(manifestListsToDelete, "manifest list");
   }
 
-  private Set<ManifestFile> readManifests(Set<Snapshot> snapshots) {
-    Set<ManifestFile> manifestFiles = Sets.newHashSet();
-    for (Snapshot snapshot : snapshots) {
-      try (CloseableIterable<ManifestFile> manifestFilesForSnapshot = readManifestFiles(snapshot)) {
-        for (ManifestFile manifestFile : manifestFilesForSnapshot) {
-          manifestFiles.add(manifestFile.copy());
-        }
-      } catch (IOException e) {
-        throw new RuntimeIOException(
-            e, "Failed to close manifest list: %s", snapshot.manifestListLocation());
-      }
+  // Returns the current set of manifests while pruning the manifests to delete
+  // If all candidate manifests are pruned the returned set of current manifests will be empty
+  private Set<ManifestFile> findCurrentManifestsAndPruneCandidates(
```

Comment:
I'm not a big fan of doing multiple things in the same helper method. In this case, I'll admit that it makes sense because we don't want to read all the manifest lists twice. Still, I find this a bit awkward. The name is long and it modifies an argument (calling `deletionCandidates.remove`). I'd probably return both sets to avoid modifying the one passed in, but that's just me. I think this is okay.

Reply:
Actually, I think there's a concurrency bug here: you don't know whether the set that was passed in is thread-safe so you shouldn't modify it. I have another solution below to consider that also fixes this.

Reply:
Yeah since it was private and we're passing in the concurrent hashset I wasn't worried too much about thread safety because it was in the method's control but agree in this case it's safer to just implement in a way that's unassuming about the input. The callback approach should address this, thanks!